### PR TITLE
[Transaction]stop TP replaying with Exception

### PR DESCRIPTION
**Pull request by liangyepianzhou**
### Motivation
When MLPendingAckStore replaying, if any ledger was deleted from bookkeeper, or ManagerLedger was fenced, MLPendingAckStore will not stop recovering and continue to report the exception.

### Modifications
End replaying when there is no ledger to read or the managerLedger is fenced.

### Verifying this change
Add a unit test.

- [x] Make sure that the change passes the CI checks.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

Check the box below and label this PR (if you have committer privilege).

Need to update docs? 

 - [x] `no-need-doc` 
 Bug Fixes.

### CODE REVIEW

**Review by congbobo184 — APPROVED** [review]

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java [hunk 1]
```diff
@@ -361,15 +360,17 @@ public void run() {
 
     class FillEntryQueueCallback implements AsyncCallbacks.ReadEntriesCallback {
 
+        private boolean isReadable = true;
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java:363]: volatile ? or AtomicBoolean

in fact outstandingReadsRequests is a AtomicLong

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java [hunk 2]
```diff
@@ -303,13 +303,12 @@ public void addFailed(ManagedLedgerException exception, Object ctx) {
...
         public void run() {
             try {
-                while (lastConfirmedEntry.compareTo(currentLoadPosition) > 0) {
-                    if (((ManagedCursorImpl) cursor).isClosed()) {
-                        log.warn("[{}] MLPendingAckStore cursor have been closed, close replay thread.",
-                                cursor.getManagedLedger().getName());
-                        return;
-                    }
-                    fillEntryQueueCallback.fillQueue();
+                if (((ManagedCursorImpl) cursor).isClosed()) {
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java:306]: this kind of casts are always a code smell and add tech debt.

can we add "isClosed" to the ManagedCursor interface ? 

otherwise if this class is meant to work only with ManagedCursorImpl then we must change the type of "cursor" to ManagedCursorImpl and to not perform this casts

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java (hunk 2 above)
Comment (liangyepianzhou) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java:306]: I think isClosed should be the basic function of ManagedCursor, so it is suitable to add an interface here. Thanks for your reminder, I will add this right away.

**Comment by liangyepianzhou** [comment]
@eolivelli please review again, THX!

**Comment by codelipenghui** [comment]
@eolivelli Please help review this PR.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java [hunk 3]
```diff
@@ -303,13 +302,12 @@ public void addFailed(ManagedLedgerException exception, Object ctx) {
...
         public void run() {
             try {
-                while (lastConfirmedEntry.compareTo(currentLoadPosition) > 0) {
-                    if (((ManagedCursorImpl) cursor).isClosed()) {
-                        log.warn("[{}] MLPendingAckStore cursor have been closed, close replay thread.",
-                                cursor.getManagedLedger().getName());
-                        return;
-                    }
-                    fillEntryQueueCallback.fillQueue();
+                if (cursor.isClosed()) {
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java:305]: why did you move this outside of the _while_ loop ?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java [hunk 4]
```diff
@@ -389,6 +389,11 @@ public Entry get() {
 
         @Override
         public void readEntriesFailed(ManagedLedgerException exception, Object ctx) {
+            if (managedLedger.getConfig().isAutoSkipNonRecoverableData()
+                    && exception instanceof ManagedLedgerException.NonRecoverableLedgerException
+                    || exception instanceof ManagedLedgerException.ManagedLedgerFencedException) {
+                isReadable = false;
+            }
             log.error("MLPendingAckStore stat reply fail!", exception);
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java:397]: is it possible to add the name of the topic ?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java (hunk 3 above)
Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java:305]: if cursor close will return ManagedledgerfenceException, so can delete this code, to speed up recovery

**Review by eolivelli — APPROVED** [review]
+1
